### ISSUE TITLE
第二轮g_loss和d_loss开始跑飞

### ISSUE THREAD

**Issue by xinxinye000** [ShawnPi233/Symbolic-Music-Genre-Transfer-with-CycleGAN-for-pytorch#1]
[epoch:   1/ 30, batch:  3060/ 3132, ite: 1021, time: 1.356423] g_loss : 7.144899, d_loss : 5.286486
[epoch:   1/ 30, batch:  3090/ 3132, ite: 1031, time: 1.336200] g_loss : 7.138608, d_loss : 5.284735
[epoch:   1/ 30, batch:  3120/ 3132, ite: 1041, time: 1.344837] g_loss : 7.103464, d_loss : 5.256520
[epoch:   2/ 30, batch:     0/ 3132, ite: 1045, time: 0.499686] g_loss : 1328.734846, d_loss : 2837.922787
[epoch:   2/ 30, batch:    30/ 3132, ite: 1055, time: 1.198439] g_loss : 10975.473752, d_loss : 23461.754297
[epoch:   2/ 30, batch:    60/ 3132, ite: 1065, time: 1.201427] g_loss : 16713.614749, d_loss : 35888.498228

情况如上，我减少了样本的数量以便更快看到结果，即使是原样本的大数据量，也是这个这个结果。

**Comment by ShawnPi233** [comment]
你好，原来的代码应该是有问题的，我后来参考其他GAN的写法，生成器和判别器分开训练就可以正常收敛了

**Comment by xinxinye000** [comment]
好的，我去试一下。



---- 回复的原邮件 ----
| 发件人 | Bingsong ***@***.***> |
| 日期 | 2024年04月22日 11:19 |
| 收件人 | ***@***.***> |
| 抄送至 | ***@***.***>***@***.***> |
| 主题 | Re: [ShawnPi233/Symbolic-Music-Genre-Transfer-with-CycleGAN-for-pytorch] 第二轮g_loss和d_loss开始跑飞 (Issue #1) |

你好，原来的代码应该是有问题的，我后来参考其他GAN的写法，生成器和判别器分开训练就可以正常收敛了

**Comment by xinxinye000** [comment]
你好，我修改代码后，还是一样的跑飞。
代码我放到附件里面了，您看是否有时间帮我看一下（我只修改了model.py文件）
[0.zip](https://github.com/ShawnPi233/Symbolic-Music-Genre-Transfer-with-CycleGAN-for-pytorch/files/15060841/0.zip)

我做了如下修改：
1、修改了归一化函数，如果不修改这个，loss一直毫无变化，原代码这里有3行代码是没有用到的。
class InstanceNorm2d(nn.InstanceNorm2d):
    def __init__(self,
                 num_features: int,
                 eps: float = 1e-5,
                 momentum: float = 0.1,
                 affine: bool = True,  # Change affine to True
                 track_running_stats: bool = False
    ) -> None:
        super(InstanceNorm2d, self).__init__(num_features, affine=affine)

    def reset_parameters(self) -> None:
        self.reset_running_stats()
        if self.affine:
            nn.init.normal_(self.weight, 1., 0.02)  # Initialize weight parameter
            nn.init.zeros_(self.bias)  # Initialize bias parameter


2、在model.py文件里面的cyclegan这个类中的生成器和鉴别器分开训练
class CycleGAN(nn.Module):
    def __init__(self, sigma=0.01, sample_size=50, lamb=10, mode='train'):
        super(CycleGAN, self).__init__()
        assert mode in ['train', 'A2B', 'B2A']
        self.G_A2B = Generator(64)
        self.G_B2A = Generator(64)
        self.D_A = Discriminator(64)
        self.D_B = Discriminator(64)
        self.D_A_all = Discriminator(64)
        self.D_B_all = Discriminator(64)
        self.l2loss = nn.MSELoss(reduction='mean')
        self.sigma = sigma
        self.mode = mode
        self.sampler = Sampler(sample_size)
        self.lamb = lamb

    def forward(self, real_A, real_B, x_m):
        # blue line
        fake_B = self.G_A2B(real_A)
        cycle_A = self.G_B2A(fake_B)

        # red line
        fake_A = self.G_B2A(real_B)
        cycle_B = self.G_A2B(fake_A)

        if self.mode == 'train':
            [sample_fake_A, sample_fake_B] = self.sampler([fake_A, fake_B])
            gauss_noise = torch.ones_like(real_A)
            gauss_noise = torch.abs(kernel_initializer(gauss_noise, mean=0, std=self.sigma))

            DA_real = self.D_A(real_A + gauss_noise)
            DB_real = self.D_B(real_B + gauss_noise)

            DA_fake = self.D_A(fake_A + gauss_noise)
            DB_fake = self.D_B(fake_B + gauss_noise)

            DA_fake_sample = self.D_A(sample_fake_A + gauss_noise)
            DB_fake_sample = self.D_B(sample_fake_B + gauss_noise)

            DA_real_all = self.D_A_all(x_m + gauss_noise)
            DB_real_all = self.D_B_all(x_m + gauss_noise)

            DA_fake_all = self.D_A_all(sample_fake_A + gauss_noise)
            DB_fake_all = self.D_B_all(sample_fake_B + gauss_noise)

            # Cycle loss
            c_loss = self.lamb * cycle_loss(real_A, cycle_A, real_B, cycle_B)

            # Generator losses
            g_A2B_loss = self.l2loss(DB_fake, torch.ones_like(DB_fake)) + c_loss
            g_B2A_loss = self.l2loss(DA_fake, torch.ones_like(DA_fake)) + c_loss

            # Discriminator losses
            d_A_loss_real = self.l2loss(DA_real, torch.ones_like(DA_real))
            d_A_loss_fake = self.l2loss(DA_fake_sample, torch.zeros_like(DA_fake_sample))
            d_A_loss = (d_A_loss_real + d_A_loss_fake) / 2
            d_B_loss_real = self.l2loss(DB_real, torch.ones_like(DB_real))
            d_B_loss_fake = self.l2loss(DB_fake_sample, torch.zeros_like(DB_fake_sample))
            d_B_loss = (d_B_loss_real + d_B_loss_fake) / 2

            # All losses
            d_A_all_loss_real = self.l2loss(DA_real_all, torch.ones_like(DA_real_all))
            d_A_all_loss_fake = self.l2loss(DA_fake_all, torch.zeros_like(DA_fake_all))
            d_A_all_loss = (d_A_all_loss_real + d_A_all_loss_fake) / 2
            d_B_all_loss_real = self.l2loss(DB_real_all, torch.ones_like(DB_real_all))
            d_B_all_loss_fake = self.l2loss(DB_fake_all, torch.zeros_like(DB_fake_all))
            d_B_all_loss = (d_B_all_loss_real + d_B_all_loss_fake) / 2

            return (c_loss, g_A2B_loss, g_B2A_loss, d_A_loss, d_B_loss,
                    d_A_all_loss, d_B_all_loss)

        elif self.mode == 'A2B':
            return fake_B, cycle_A
        elif self.mode == 'B2A':
            return fake_A, cycle_B

**Comment by Jesseill** [comment]
> 你好，原来的代码应该是有问题的，我后来参考其他GAN的写法，生成器和判别器分开训练就可以正常收敛了

您好，請問可以分享您改的code嗎? 我使用原本的Dataset也遇到 epoch 2 之後loss起飛的問題。謝謝！
